client/state/themes/actions.js

```diff
@@ -116,33 +116,49 @@ export function receiveThemes( themes, siteId ) {
 /**
  * Triggers a network request to fetch themes for the specified site and query.
  *
- * @param  {Number|String} siteId Jetpack site ID or 'wpcom' for any WPCOM site
- * @param  {String}        query  Theme query
- * @return {Function}             Action thunk
+ * @param  {Number|String} siteId        Jetpack site ID or 'wpcom' for any WPCOM site
+ * @param  {Object}        query         Theme query
+ * @param  {String}        query.search  Search string
+ * @param  {String}        query.tier    Theme tier: 'free', 'premium', or '' (either)
+ * @param  {String}        query.filter  Filter
+ * @param  {Number}        query.number  How many themes to return per page
+ * @param  {Number}        query.offset  At which item to start the set of returned themes
+ * @param  {Number}        query.page    Which page of matching themes to return
+ * @return {Function}                    Action thunk
  */
 export function requestThemes( siteId, query = {} ) {
 	return ( dispatch, getState ) => {
 		const startTime = new Date().getTime();
-		let siteIdToQuery, queryWithApiVersion;
-
-		if ( siteId === 'wpcom' ) {
-			siteIdToQuery = null;
-			queryWithApiVersion = { ...query, apiVersion: '1.2' };
-		} else {
-			siteIdToQuery = siteId;
-			queryWithApiVersion = { ...query, apiVersion: '1' };
-		}
 
 		dispatch( {
 			type: THEMES_REQUEST,
 			siteId,
 			query
 		} );
 
-		return wpcom.undocumented().themes( siteIdToQuery, queryWithApiVersion ).then( ( { found, themes: rawThemes } ) => {
+		let request;
+
+		if ( siteId === 'wporg' ) {
+			request = () => wporg.fetchThemesList( query );
+		} else if ( siteId === 'wpcom' ) {
+			request = () => wpcom.undocumented().themes( null, { ...query, apiVersion: '1.2' } );
+		} else {
+			request = () => wpcom.undocumented().themes( siteId, { ...query, apiVersion: '1' } );
+		}
+
+		// WP.com returns the number of results in a `found` attr, so we can use that right away.
+		// WP.org returns an `info` object containing a `results` number, so we destructure that
+		// and use it as default value for `found`.
+		return request().then( ( { themes: rawThemes, info: { results } = {}, found = results } ) => {
 			let themes;
 			let filteredThemes;
-			if ( siteId !== 'wpcom' ) {
+			if ( siteId === 'wporg' ) {
+				themes = map( rawThemes, normalizeWporgTheme );
```

Comment:
Wondering if it might be tidier here to use a table of normalizing functions keyed by source. That would just leave the special case for jetpack filtering, rather than a three-way `if/else`.